backend/api/src/realms.rs

```diff
@@ -84,14 +110,59 @@ impl Tree {
             }
         }
 
-        Ok(Tree { realms, children })
+        // We also need a map from the full path to the proper realm,
+        // and conversely, we want to cache the full path inside the realm.
+        let from_path = index_by_path(&mut realms, &children)?;
+
+        fn index_by_path(
+            realms: &mut HashMap<u64, Realm>,
+            children: &HashMap<u64, Vec<u64>>,
+        ) -> anyhow::Result<HashMap<String, u64>> {
+
+            let mut index = HashMap::new();
+
+            index.insert("".to_string(), 0);
+
+            recurse(0, realms, children, &mut index)?;
+
+            fn recurse(
+                parent: u64,
+                realms: &mut HashMap<u64, Realm>,
+                children: &HashMap<u64, Vec<u64>>,
+                index: &mut HashMap<String, u64>,
+            ) -> anyhow::Result<()> {
+
+                let parent_path = realms.get(&parent).context("realm structure invalid")?.path.clone();
+
+                if let Some(current_children) = children.get(&parent) {
+                    for child_id in current_children {
+
+                        let child = realms.get_mut(child_id).context("realm structure invalid")?;
+                        child.path = format!("{}/{}", parent_path, child.path);
+                        index.insert(child.path.clone(), *child_id);
+
+                        recurse(*child_id, realms, children, index)?;
+                    }
+                }
+
+                Ok(())
+            }
+
+            Ok(index)
+        }
+
+        Ok(Tree { realms, children, from_path })
     }
 
     pub(crate) fn get_node(&self, id: &Id) -> Option<&Realm> {
         self.realms.get(&id.key_for(Id::REALM_KIND)?)
     }
 
     pub(crate) fn root(&self) -> &Realm {
-        self.realms.get(&0).expect("bug: no root realm")
+        self.realms.get(&0).unwrap()
```

Comment:
We really need to figure out where we once make sure that the whole tree is valid. And then assume that it's correct everywhere else.

Either way, `get().unwrap()` should always be:

```suggestion
        self.realms[&0]
```

Reply:
Yeah good point. I added a comment to that effect in the code. I think the point in question is after having put all the child keys in place.

However, it's still a valid question, I think, whether or not we just want to crash when the structure is invalid because of a bug. I would be inclined to say "yes", but somethingsomething high availability.

Re: `get(...).unwrap()`: There should be none left.

Reply:
Well, with #79, panics are "properly" handled. So this is just a question whether we want 500 or a graphql error.